packages/nimble-components/src/table/specs/table-column-specs/table-column-mapping.md

```diff
@@ -0,0 +1,359 @@
+        Archived
+        <nimble-mapping-icon key="true" icon="nimble-icon-database" label="Archived"></nimble-mapping-icon>
+    </nimble-table-column-icon>
+</nimble-table>
+```
+
+Each column contains mapping elements that define what to render when the cell's value matches the given `key` value.
+
+When none of the given mappings match the record value for a cell, that cell will be empty. Alternatively, if one of the mappings has the `default` attribute, it will match when no other mappings have. This is equivalent to the `placeholder` configuration we provide on `nimble-table-column-text` and `nimble-table-column-anchor`.
+
+Validation will be performed to ensure each mapping's key value can be converted to the `data-type` of the column. If not, an error flag will be set on the column's validation object. Note that whenever an error flag is set on the column's validation object, a generic `invalidColumnConfiguration` flag is also set on the table, putting it in an invalid state as well.
+
+If multiple mappings in a column have the same key, an error flag will be set on the column's validity object.
+
+If an invalid `icon` value is passed to `nimble-mapping-icon`, an error flag will be set on the column's validity object. An invalid `icon` value is any element that cannot be resolved or that does not derive from `Icon`.
+
+`nimble-table-column-icon` supports only `nimble-mapping-icon` and `nimble-mapping-spinner` as mapping elements. `nimble-table-column-mapping` supports only `nimble-mapping-text`. Unsupported mappings will result in an error flag being set on the column's validity object.
+
+Text in a grouping header or in the cell will be ellipsized and gain a tooltip when the full text is too long to display.
+
+**Alternatives:**
+
+An earlier version of this spec proposed mapping elements with `template` elements as content instead of relying on attribute configuration. The template element would define the mapped html to render. We would impose restrictions on the types of supported elements that could be provided in the template.
+
+Pros:
+
+-   It would not require updates to the API if we needed to support new types of mapped content (e.g. icon with text), or if the mapped content itself got new configuration options (e.g. a scaling factor for icons).
+-   Undefined element types caught at compile time.
+
+Cons:
+
+-   Verbose. Requires user to create `template` element and wrap text in `span`s for styling purposes.
+-   Requires difficult validation to ensure only supported elements are present in the `template`.
+-   Could allow users to provide inline styling.
+-   Blazor: cannot put Blazor components inside `template`--must use raw Nimble elements without type safety
+
+### API
+
+#### Icon column element:
+
+_Component Name_
+
+-   `nimble-table-column-icon`
+
+_Props/Attrs_
+
+-   `field-name`: string
+-   `data-type`: 'string' | 'number' | 'boolean' | null (defaults to 'string')
```

Comment:
What would a `null` value for `data-type` mean? If it always results in an error, is it possible to disallow that value?

Reply:
In the current spec it shows a null type referring to a type of string, but also that "string" is a possible value.
I think we should only have one representation of each format, either:
 - undefined/null (default: string) / "number" / "boolean" (2 keys and a default)
 - "string" / "number" / "boolean" and no default value so error if null /undefined (3 keys and no default)

I don't feel strongly about which of those options we choose. In general I prefer good defaults (the 2 keys option) but in this case maybe choosing a type by default could lead to subtle bugs. For example I forget to specify the type and pass numbers as strings which are then treated as strings in comparisons of "1" and "1.0" and "+1.0" and "1e0", etc.

So in this case I think I prefer 3 keys no default, i.e. where Jesse is going.

Reply:
Good point, it's a tradeoff between:
- 2 keys and a default being easier to configure (many cases won't need to specify `data-type` at all)
- 3 keys and no default being less likely to introduce subtle bugs because clients must always specify `data-type` so presumably they'd also think about its value

I'm good with 3 keys no default.

Reply:
Updated spec